Migration to uv & Nixify using python-flake

semi710 opened the pull request:
**Resolves:** #3

### TODO

* [x] Expose `code_as_data` Python module as a `.#default` Nix package
* [x] Add `devShell` for development environment
* [x] Support `wheel` distribution via `nix run .#wheel`
* [x] Add `test` target runnable via `nix run .#test`
* [x] Add `ci` using [`omnix`](https://github.com/juspay/omnix)
* [x] Migrate Postgres scripts to [`services-flake`](https://github.com/juspay/services-flake)
* [ ] Addressed all Feedback from @eswar2001 
* [ ] Update README with:
  * [ ] Relevant changes
  * [ ] Explanation of what/how things work with `uv` and `nix`
* [ ] Squash, and ensure the commit is following the pattern used in repo's prev commits.
* [ ] Final review from @srid

semi710 committed:
chore: Migration to uv

semi710 committed:
chore: Nixify using uv2nix

tests can be run using `nix run .#test` and `wheel` dist can be build
using `nix build .#wheel`.

`default` exposes `code_as_data` python module.

semi710 committed:
chore: exposed package fixed

The exposed `default` pacakge requires same python as `code-as-data`
repo.

The `default` pacakge can be used inside python packages and it will
work.

semi710 committed:
chore: Add service-flake for postgres

semi710 commented:
@eswar2001 The Pr is near completion please check for anything missing or doing an unintentional behaviour?

- `nix develop` will get you in development shell. (`direnv allow` will make this automatic when doing `cd` in the dir)
	- The development environment is now migrated to use `uv` for package management.
	- Addition of package can be done by doing `uv add <packagename>`.
	- uv build will build `whl` file for the module but nix way is recommended.	
- `nix run .#wheel` will build the `whl` file and place it in `result` dir though nix exposes the module as a `default` package so it can be added in any python env like the snippet below.
```nix
(python311.withPackages (p: with p; [
  websockets
  aiohttp
  boto3
  inputs.code-as-data.${pkgs.system}.default
]))

```

### Service
> `postgres` service is now switched to [`services-flake`](https://github.com/juspay/services-flake)

to run the service do 
```sh
# -t flag will disable tui mode (you can remove the flag if want tui)
nix run .#code-as-data -- -t=false 
```

I have tested the service by running `scripts/setup_db.py` file and it works.

>[!NOTE] 
> Currently the service has port hard code.

My question regarding `.env` is that is this file used by any python file or something? if so then this can be generated by nix also.
i mean i was not able to purely understand the use of the .env file coz the port is picked up randomly too https://github.com/juspay/code-as-data/blob/c573e85c397e11f41c9f01a29717543bb70017b4/flake.nix#L78

If you can shed light on the req of `.env` file in this situation i can do the changes with nix reqiured.
else if the file is not explicitly needed we can define these env variables in devShell itself.

eswar2001 commented:
> @eswar2001 The Pr is near completion please check for anything missing or doing an unintentional behaviour?
> 
> * `nix develop` will get you in development shell. (`direnv allow` will make this automatic when doing `cd` in the dir)
>   
>   * The development environment is now migrated to use `uv` for package management.
>   * Addition of package can be done by doing `uv add <packagename>`.
>   * uv build will build `whl` file for the module but nix way is recommended.
> * `nix run .#wheel` will build the `whl` file and place it in `result` dir though nix exposes the module as a `default` package so it can be added in any python env like the snippet below.
> 
> ```nix
> (python311.withPackages (p: with p; [
>   websockets
>   aiohttp
>   boto3
>   inputs.code-as-data.${pkgs.system}.default
> ]))
> ```
> 
> ### Service
> > `postgres` service is now switched to [`services-flake`](https://github.com/juspay/services-flake)
> 
> to run the service do
> 
> ```shell
> # -t flag will disable tui mode (you can remove the flag if want tui)
> nix run .#code-as-data -- -t=false 
> ```
> 
> I have tested the service by running `scripts/setup_db.py` file and it works.
> 
> Note
> 
> Currently the service has port hard code.
> 
> My question regarding `.env` is that is this file used by any python file or something? if so then this can be generated by nix also. i mean i was not able to purely understand the use of the .env file coz the port is picked up randomly too
> 
> https://github.com/juspay/code-as-data/blob/c573e85c397e11f41c9f01a29717543bb70017b4/flake.nix#L78
> 
> If you can shed light on the req of `.env` file in this situation i can do the changes with nix reqiured. else if the file is not explicitly needed we can define these env variables in devShell itself.

yeah , if nix could update it as .env then python could read it directly from there or we can also try moving to exporting to ENV in devshell

everything looks good , I'll try it out . also in service-flake is it possible to say use n cores for a specific service ?

semi710 committed:
ci: Add omnix ci

semi710 committed:
chore: Read `.env` file to load env variables in `devShell.default`

semi710 commented:
Sorry mate, not sure how, but Git didn't notify me about your comment.

> **service-flake** – Is it possible to specify using *n* cores for a specific service?

How are you currently handling this?

> If Nix could update it as `.env`...

So in `devShells.nix`, we're now reading the `.env` file to load environment variables into the devShell.
And in `services.nix`, we're using those variables (like `PORT`) directly from there.

---

@eswar2001 Please test this and share your feedback.
Once confirmed, the README can be updated with a `nix` section reflecting the relevant changes.
After that, I’ll squash all the commits, and the PR will be ready to merge.

srid commented on .envrc:
```diff
@@ -0,0 +1 @@
+use flake
```
We should use `om develop` across all (new) projects, and ideally existing projects as well.

srid commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,34 @@
+          - runner: ubuntu-latest
+            system: x86_64-linux
+          - runner: macos-latest
+            system: aarch64-darwin
```
All new OSS projects on GitHub should use our self-hosted runners.

srid commented on .github/workflows/ci.yaml:
```diff
@@ -0,0 +1,34 @@
+        uses: DeterminateSystems/nix-installer-action@main
+      - name: "Install omnix"
+        run: |
+          nix profile install nixpkgs#omnix
```
These steps become redundant when we use self-hosted runners.

srid commented on nix/devShells.nix:
`direnv` can read `.env` file; I don't think we should do the plumbing manually in Nix unless there's a reason not to use direnv.

srid replied:
Besides, we should try to avoid low-level implementation wherever possible.

srid replied:
@eswar2001 If `.env` is used only in devShell (and not in Nix package, à la `nix run ..`), then would direnv's dotenv be sufficient for this project's developers? I assume they would use `direnv` over `nix develop` in that case.

See https://nixos.asia/en/direnv

semi710 replied:
> `direnv` can read `.env` file; I don't think we should do the plumbing manually

Yep, this way we have a very clean code too.
Though, user/developer need to be aware that `nix run` without direnv env activated will throw error; so `direnv allow` is a mandate.

---

But @srid by doing this way; I am able to get the variable value in the `repl` but `nix run` is unaware of that variable.

As in `services-flake` 
```nix
port = lib.toInt (builtins.getEnv "DB_PORT");
```
in repl it gives the output of correct port as integer, But while doing a `nix run` the Value of port errors out because for `nix run` `DB_PORT` is "".

srid commented on nix/python.nix:
Add a comment to this module, saying the bulk of this will be superceded by `python-flake` (with a link).

srid reviewed (commented):
Just thought I'd do an initial review of the draft. Will be happy to do a final thorough review once the PR is ready.

semi710 committed:
chore(nix): omnix in use

semi710 committed:
chore(nix): review comments

semi710 committed:
ci: switch to self hosted runners

semi710 committed:
chore(nix): Add env variables

semi710 committed:
chore(nix): remove `om.yaml` entry from .envrc

semi710 committed:
just: Add justfile with recipes

semi710 committed:
refactor: Migrate to `python-flake`

Python flake uses `uv2nix` under the hood and exposes api to develop
python envs easily.

semi710 committed:
omnixrc: udpate omnixrc url in `.envrc`

semi710 committed:
justfile: Add relevant comments

semi710 committed:
omnixrc: udpated to clean-url

srid commented:
@niksingh710  Now that it is using python-flake, please update the title & description of the PR. And mark it as draft.
